File: src/02.b.type_classes.ts
type ToString<A> = {
  stringify: (a: A) => string
}

const boolToString: ToString<boolean> = {
  stringify: (a) => a ? 'true' : 'false'
}

const boolToStringCapitalized: ToString<boolean> = {
  stringify: (a) => a ? 'TRUE' : 'FALSE'
}

boolToString.stringify(true)
boolToString.stringify(false)

type Foo = { tag: 'Bar' } | { tag: 'Baz' }

const fooToString: ToString<Foo> = {
  stringify: (a) => a.tag
}

fooToString.stringify({ tag: 'Bar' })

const greetings = <A>(toString: ToString<A>) => (x: A) =>
  `Hello ${toString.stringify(x)}!!!`

greetings(fooToString)({ tag: 'Bar' })

greetings(boolToString)(true)
greetings(boolToString)(false)

greetings(boolToStringCapitalized)(false)

File: src/02.b.type_classes.hs
-- TypeClasses

class ToString a where
  stringify :: a -> String

instance ToString Bool where
  stringify x =
    if x
      then "True"
      else "False"

data Foo = Bar | Baz

instance ToString Foo where
  stringify x =
    case x of
      Bar -> "Bar"
      Baz -> "Baz"

greetings :: ToString a => a -> String
greetings x = "Hello " ++ stringify x ++ "!!!"

main = do
  putStrLn (stringify True)
  putStrLn (stringify False)

  putStrLn ("Hola " ++ stringify False ++ " " ++ stringify True)

  putStrLn (stringify True)

  putStrLn (stringify Bar)
  putStrLn (stringify Baz)

  putStrLn (greetings True)
  putStrLn (greetings Bar)
